### PR TITLE
bpo-28598: Support __rmod__ for RHS subclasses of str in % string formatting operations

### PR DESCRIPTION
**Pull request by mjpieters**
When you use `'%s' % SubClassOfStr()`, where `SubClassOfStr.__rmod__` exists, the reverse operation is ignored as normally such string formatting operations use the `PyUnicode_Format()` fast path. This patch tests for subclasses of `str` first and picks the slow path in that case.

Fixes bpo-28598.

### CODE REVIEW

**Comment by mjpieters** [comment]
Open question: Should this patch be backported to 3.6 and 3.5? It's a bug fix, not a new feature.

**Comment by berkerpeksag** [comment]
> Should this patch be backported to 3.6 and 3.5?

Yes, it should be backported if it's a bug (I haven't investigated myself yet.)

> I'm not sure if issues are linked automatically yet.

We don't use `issueNNNN` anymore. The correct way to link PRs to bugs.p.o is `bpo-NNNN`.

**Comment by mjpieters** [comment]
> We don't use issueNNNN anymore. The correct way to link PRs to bugs.p.o is bpo-NNNN.

Ta, also for the title change! Updated the 'fixes' line too.

File: Python/ceval.c [hunk 1]
```diff
@@ -1355,9 +1355,15 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
         TARGET(BINARY_MODULO) {
             PyObject *divisor = POP();
             PyObject *dividend = TOP();
-            PyObject *res = PyUnicode_CheckExact(dividend) ?
-                PyUnicode_Format(dividend, divisor) :
-                PyNumber_Remainder(dividend, divisor);
+            PyObject *res;
+            if (PyUnicode_CheckExact(dividend) && !(
+                  PyUnicode_Check(divisor) && !PyUnicode_CheckExact(divisor))) {
```
Comment (serhiy-storchaka) [inline Python/ceval.c]: Too many negations.

I would expand the expression as `PyUnicode_CheckExact(dividend) && (!PyUnicode_Check(divisor) || PyUnicode_CheckExact(divisor))`.

File: Misc/NEWS [hunk 2]
```diff
@@ -10,6 +10,8 @@ What's New in Python 3.7.0 alpha 1?
 Core and Builtins
 -----------------
 
+- Issue #28598: Support __rmod__ for subclasses of str being called before str.__mod__.
```
Comment (serhiy-storchaka) [inline Misc/NEWS]: Add "Patch by Martijn Pieters."

**Review by serhiy-storchaka — COMMENTED** [review]
In general LGTM except few minor details.

And I'm not sure about the style of commit messages now. Should it include issue number at the first line? Should it contain "Patch by YYY." (now the author is specified in metadata)?

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by brettcannon** [comment]
Should this really be backported? It seems like a change in semantics more than a bugfix.

**Comment by mjpieters** [comment]
@brettcannon This is definitely a bug, not a change in semantics. From the [`object.__rmod__` documentation](https://docs.python.org/3/reference/datamodel.html#object.__rmod__):

> **Note**: If the right operand’s type is a subclass of the left operand’s type and that subclass provides the reflected method for the operation, this method will be called before the left operand’s non-reflected method. This behavior allows subclasses to override their ancestors’ operations.

`str % subclass_of_str` doesn't follow this behavior, and `str` is the only type that doesn't (everything else goes through the `PyNumber_Remainder()` slow path). This patch fixes that bug.

**Pull request merged**